src/XMakeBuildEngine/Resources/Constants.cs

```diff
@@ -219,43 +219,94 @@ internal static class AvailableStaticMethods
         /// </summary>
         private static Object s_locker = new Object();
 
+        static AvailableStaticMethods()
+        {
+            InitializeAvailableMethods();
+        }
+
         /// <summary>
         /// Whether a key is present
         /// </summary>
         internal static bool ContainsKey(string key)
         {
-            InitializeAvailableMethods();
-
             return s_availableStaticMethods.ContainsKey(key);
         }
         /// <summary>
         /// Add an entry if not already present
         /// </summary>
         internal static bool TryAdd(string key, Tuple<string, Type> value)
         {
-            InitializeAvailableMethods();
-
             return s_availableStaticMethods.TryAdd(key, value);
         }
 
         /// <summary>
+        /// Constructs the fully qualified method name (<see cref="typeFullName"/>::<see cref="simpleMethodName"/> and adds it to the cache)
+        /// </summary>
+        /// <param name="typeFullName"></param>
+        /// <param name="simpleMethodName"></param>
+        /// <param name="typeInformation"></param>
+        /// <returns></returns>
+        public static bool TryAdd(string typeFullName, string simpleMethodName, Tuple<string, Type> typeInformation)
+        {
+            return TryAdd(CreateQualifiedMethodName(typeFullName, simpleMethodName), typeInformation);
+        }
+
+        /// <summary>
         /// Get an entry if present
         /// </summary>
         internal static bool TryGetValue(string key, out Tuple<string, Type> value)
         {
-            InitializeAvailableMethods();
-
             return s_availableStaticMethods.TryGetValue(key, out value);
         }
 
         /// <summary>
+        /// Get an entry or null if not present
+        /// </summary>
+        internal static Tuple<string, Type> GetValue(string key)
+        {
+            Tuple<string, Type> typeInformation;
+            return s_availableStaticMethods.TryGetValue(key, out typeInformation) ? typeInformation : null;
+        }
+
+        /// <summary>
+        /// Tries to retrieve the type information for a type name / method name combination. 
+        /// 
+        /// It does 2 lookups:
+        /// 1st try: <see cref="typeFullName"/>;
+        /// 2nd try: <see cref="typeFullName"/>::<see cref="simpleMethodName"/>
+        /// 
+        /// </summary>
+        /// <param name="typeFullName">namespace qualified type name</param>
+        /// <param name="simpleMethodName"> name of the method</param>
+        /// <returns></returns>
+        internal static Tuple<string, Type> GetTypeInformationFromTypeCache(string typeFullName, string simpleMethodName)
+        {
+            var methodFullName = CreateQualifiedMethodName(typeFullName, simpleMethodName);
```

Comment:
Should this be inlined so we don't have to call `CreateQualifiedMethodName` if the `typeFullName` works?